doc/design/scope.md

```diff
@@ -0,0 +1,58 @@
+# Scope
+
+### Define
+
+Scope is a context to manage Variables. It mainly contains a map from Variable name to Variable. Net will get and update variable throw scope.
+
+```cpp
+class Variable;
+using VariablePtr = std::shared_ptr<Variable>;
```

Comment:
Maybe we should have a global header called `TypeDefs.h`. In this header, we specify all default types for `Map`, `String`, `Set`, etc. Like

```cpp
namespace paddle {
template <typename KEY, typename VAL>
using Map = std::unordered_map<KEY, VAL>;
using String = std::string;
template <typename T>
using Set = std::unordered_set<T>;
}
```